### PR TITLE
Serialize anonymous functions (fixes #11)

### LINKED ISSUE #11: possibly read/write functions?
I have encountered a situation in which i have a array of anonymous functions that i want to save for later use.

but for now, the JLD complained that ```it cannot write a pointer to JLD file```.

I am curious that is it possible to do this, may be by save the function as an expression in some form of string and re-eval when read into current work space.

any thoughts?

### PR DESCRIPTION
**Pull request by amack315**
This pull request resolves #11 by adding serialization for anonymous functions via lowered ast expressions.

### CODE REVIEW

File: src/JLD.jl [hunk 1]
```diff
@@ -767,6 +767,21 @@ if VERSION >= v"0.4.0-dev+4319"
     writeas(x::SimpleVector) = SimpleVectorWrapper([x...])
 end
 
+# Serializer for anonymous functions
+# convert functions to lowered ast expressions
+function func2expr(fun::Function)
+    @assert !isa(fun.env, MethodTable) "generic functions not supported"
```
Comment (timholy) [inline src/JLD.jl]: Replace with `isgeneric(fun)`?

**Comment by timholy** [comment]
This is very nice!

I can't replicate the test failure on current master, but if I pull your branch I can reproduce the failure. Consequently, it's a consequence of this PR, but at a level that "should work" but doesn't. Does this pass locally for you on 0.4? (If you have a 0.4 installation.) If so, what version of julia are you running? We could git-bisect to see what changed.

**Comment by amack315** [comment]
I didn't test in 0.4 beforehand. It looks like in 0.4, ASTs use a new type, GlobalRef, to refer to names in the function using a global module path, which was causing the test failure. To fix this, I've added a GlobalRefSerializer. 

The GlobalRefSerializer stores names relative to the workspace of the current module, rather than in a global manner, so that when deserialized the names should still work even if the current module is in a different spot in the module hierarchy (or even if it's a different module entirely). For example, if the current module (the module using JLD) is `MyMod`, and the function refers to `MyMod.SubMod.var`, then`MyMod.Submod.var` gets saved as`Submod.var`. I think this makes sense since (in normal uses of the language), there's no way to directly reference modules that are not in the workspace of the current module, so there's no need to reference the absolute module paths found in GlobalRef objects.

**Comment by amack315** [comment]
It looks like some tests are still failing on Travis CI. However, they pass for me on v"0.4.0-dev+6807".

**Comment by amack315** [comment]
Let me clarify my last comment. The reason why the test was failing was because function asts in v0.4 use GlobalRef objects to refer to names. Each GlobalRef object contains a field called mod, which is of type Module, and a field called name, which is a Symbol. JLD was getting tripped up when trying to serialize the module.

I wrote a GlobalRef serializer last night that stores the mod field of a GlobalRef as a Tuple of symbols, using relative module paths (relative to current_module() which I think may not be the most sensible choice). However, looking at the docs, it looks like this choice is inconsistent with the rest of JLD, in particular how JLD handles types. So I plan on changing GlobalRefSerializer so that the full module path is stored by default. I will also look into using truncate_module_paths() and/or rootmodule to allow users to store functions/GlobalRefs with relative module paths.

**Comment by amack315** [comment]
I've added commits so that anonymous functions and GlobalRefs get serialized using absolute module paths. I think adding the option to serialize anonymous functions using relative module paths is best left for another pull request. So, while I had some issues with the logic of the first version of this pull request, I'm happy with the logic of the current version, and think it's ready for feedback from others in the project.

However, it looks like there are still test issues, given the results from Travis-CI. This is puzzling to me, as when I run Pkg.test("JLD") on my branch locally, the test passes, as long as I don't run the DataFrames test (i.e., I comment out `runtest("jld_dataframe.jl")` in runtests.jl). I get the same results when I run Pkg.test("JLD") on master, (i.e. the test passes when I comment out the DataFrame test, but fails with the DataFrame test). However, looking at the log from Travis CI, it looks like it might be running a more involved test than what Pkg.test() does.

**Comment by timholy** [comment]
Sorry for being pretty inattentive here, I've been just swamped with https://github.com/JuliaLang/Color.jl/issues/101.

I'm really impressed with your progress! Regarding the Travis failure, I suspect that what's happening is that it's failing on the JLDArchives tests: https://github.com/JuliaLang/JLD.jl/blob/033ad195794b2cd5e42cc90229ce85303bdbe82c/.travis.yml#L13. I just fixed that in #17.

File: test/jldtests.jl [hunk 2]
```diff
@@ -448,6 +507,10 @@ for compress in (true,false)
         @check fidr d
         exr = read(fidr, "ex")   # line numbers are stripped, don't expect equality
         checkexpr(ex, exr)
+        funr = read(fidr, "fun")
+        checkfuns(fun, funr)
+        function_referencing_module_r = read(fidr, "function_referencing_module")
+        checkfuns(function_referencing_module, function_referencing_module_r)
```
Comment (timholy) [inline test/jldtests.jl]: All of this is really nicely done.

Unless I'm missing something, you don't test that you can call a function that you load from a file. While it's trivial (you've checked the ASTs very carefully), it seems like a good sanity check.

**Comment by timholy** [comment]
I'll go ahead and merge this, and add that check myself. Thanks for the great contribution!

**Comment by timholy** [comment]
See https://github.com/JuliaLang/JLD.jl/commit/3d88c03e0bdd06db8208f419f85b4f37b2ac41cd?w=1

**Comment by amack315** [comment]
Sorry for the late reply, I was traveling the last few days. I agree it makes sense to test you can call a function loaded from a file. Thanks for adding it!

**Pull request merged**